pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/MaximumPinotConsumptionDelayTracker.java

```diff
@@ -0,0 +1,306 @@
+    return _realTimeTableDataManager.isPartitionHostedInThisServer(partitionId);
+  }
+
+  /*
+   * Helper function to be called when we should stop tracking a given partition. Removes the partition from
+   * all our maps, it also updates the maximum if the tracked partition was the previous maximum.
+   *
+   * @param partitionId partition ID which we should stop tracking.
+   */
+  private void removePartitionId(long partitionId) {
+    _partitionToDelaySampleMap.remove(partitionId);
+    // If we are removing a partition we should stop reading its ideal state.
+    _partitionsMarkedForConfirmation.remove(partitionId);
+   }
+
+  // Custom Constructor
+  public MaximumPinotConsumptionDelayTracker(ServerMetrics sm, String tableNameWithType, RealtimeTableDataManager tdm,
+      int maxNumberOfTicksBeforeVerification, int initialTimeThreadDelayMs, int timerThreadTickIntervalMs)
+      throws RuntimeException {
+    _serverMetrics = sm;
+    _tableNameWithType = tableNameWithType;
+    _realTimeTableDataManager = tdm;
+    // TBD handle negative timer values
+    if ((maxNumberOfTicksBeforeVerification < 0) || (initialTimeThreadDelayMs < 0)
+        || (timerThreadTickIntervalMs <= 0)) {
+      throw new RuntimeException("Illegal timer arguments");
+    }
+    _enableAging = true;
+    _maxNumberOfTicksBeforeVerification = maxNumberOfTicksBeforeVerification;
+    _initialTimeThreadDelayMs = initialTimeThreadDelayMs;
+    _timerThreadTickIntervalMs = timerThreadTickIntervalMs;
+    _timerTask = new TrackingTimerTask("TimerTask" + tableNameWithType, this);
+    _timer = new Timer();
+    _timer.schedule(_timerTask, _initialTimeThreadDelayMs, _timerThreadTickIntervalMs);
+    // Install callback metric
+    _serverMetrics.addCallbackTableGaugeIfNeeded(_tableNameWithType, ServerGauge.MAX_PINOT_CONSUMPTION_DELAY_MS,
+        () -> (long) getMaxPinotConsumptionDelay());
+  }
+
+  // Constructor that uses default timeout
+  public MaximumPinotConsumptionDelayTracker(ServerMetrics sm, String tableNameWithType,
+      RealtimeTableDataManager tdm) {
+    this(sm, tableNameWithType, tdm, MAX_NUMBER_OF_TIME_TICKS_BEFORE_VERIFICATION,
+        INITIAL_TIMER_THREAD_DELAY_MS, TIMER_THREAD_TICK_INTERVAL_MS);
+  }
+
+  /**
+   * Use to set or rest the aging of reported values.
+   * @param enableAging true if we want maximum to be aged as per sample time or false if we do not want to age
+   *                   samples
+   */
+  @VisibleForTesting
+  void setEnableAging(boolean enableAging) {
+    _enableAging = enableAging;
+  }
+
+  /*
+   * Called by LLRealTimeSegmentDataManagers to post delay updates to this tracker class.
+   * If the new sample represents a new Maximum we update the current maximum.
+   * If the new sample was for the partition that was maximum, but delay is not maximum anymore, we must select
+   * a new maximum.
+   *
+   * @param delayInMilliseconds pinot consumption delay being recorded.
+   * @param currentTime sample time.
```

Comment:
call it sampleTime then? If currenttime, then you can call System.currentMillis() in this code, and avoid the parameter

Reply:
Renamed